pkg/ebpf/c/tracee.bpf.c

```diff
@@ -4315,51 +4303,54 @@ int BPF_KPROBE(cgroup_bpf_run_filter_skb)
+    //              sometimes is zeroed. In this case, we use the skb data
+    //              pointer as a second key for a second entry in the map
+    //              containing the same data.
     //
     // Theoretically, map collisions might occur, BUT very unlikely due to:
     //
     // kprobe (map update) -> cgroup/skb (consume) -> kretprobe (map delete)
+    //
+    // Tests using localhost traffic show that this method causes very little
+    // packet loss. In fact, there was a single packet loss in the test bellow
+    // (out of 11.8 GBytes transferred), which is a 0.000008% packet loss.
+    //
+    // [  5] local 127.0.0.1 port 33388 connected to 127.0.0.1 port 5201
+    // [ ID] Interval           Transfer     Bitrate         Retr  Cwnd
+    // [  5]   0.00-1.00   sec  1.25 GBytes  10.8 Gbits/sec    0    895 KBytes
+    // [  5]   1.00-2.00   sec  1.23 GBytes  10.6 Gbits/sec    0   1.75 MBytes
+    // [  5]   2.00-3.00   sec  1.26 GBytes  10.9 Gbits/sec    0   1.75 MBytes
+    // [  5]   3.00-4.00   sec  1.25 GBytes  10.8 Gbits/sec    0   1.75 MBytes
+    // [  5]   4.00-5.00   sec  1.31 GBytes  11.2 Gbits/sec    0   1.75 MBytes
+    // [  5]   5.00-6.00   sec   998 MBytes  8.38 Gbits/sec    0   1.75 MBytes
+    // [  5]   6.00-7.00   sec  1.08 GBytes  9.26 Gbits/sec    0   1.75 MBytes
+    // [  5]   7.00-8.00   sec  1.01 GBytes  8.68 Gbits/sec    0   1.75 MBytes
+    // [  5]   8.00-9.00   sec  1.33 GBytes  11.5 Gbits/sec    0   1.75 MBytes
+    // [  5]   9.00-10.00  sec  1.09 GBytes  9.32 Gbits/sec    0   1.81 MBytes
+    // - - - - - - - - - - - - - - - - - - - - - - - - -
+    // [ ID] Interval           Transfer     Bitrate         Retr
+    // [  5]   0.00-10.00  sec  11.8 GBytes  10.1 Gbits/sec    0   sender
+    // [  5]   0.00-10.00  sec  11.8 GBytes  10.1 Gbits/sec        receiver
+    //
 
-    bpf_map_update_elem(&cgrpctxmap, &skbts, &neteventctx, BPF_NOEXIST);
-
-    return 0;
-}
-
-SEC("kretprobe/__cgroup_bpf_run_filter_skb")
-int BPF_KRETPROBE(ret_cgroup_bpf_run_filter_skb)
-{
-    // runs AFTER the CGROUP/SKB eBPF program
-
-    // pick from entry from entrymap
-    u32 host_tid = bpf_get_current_pid_tgid();
-    struct entry *entry = bpf_map_lookup_elem(&entrymap, &host_tid);
-    if (!entry) // no entry == no tracing
-        return 0;
-
-    // pick args from entry point's entry
-    // struct sock *sk = (void *) entry->args[0];
-    struct sk_buff *skb = (void *) entry->args[1];
-
-    // cleanup entrymap
-    bpf_map_delete_elem(&entrymap, &host_tid);
-
-    // use skb timestamp as the key for cgroup/skb
-    u64 skbts = BPF_READ(skb, tstamp);
+    if (skbts) // first attempt with skb timestamp
+        bpf_map_update_elem(&cgrpctxmap, &skbts, &neteventctx, BPF_NOEXIST);
 
-    // delete netctx after cgroup ebpf program runs
-    bpf_map_delete_elem(&cgrpctxmap, &skbts);
+    if (data) // second attempt will be with skb data pointer
```

Comment:
There is no termination after the first attempt, by design? Or are the field mutually exclusive on being NULL (if `data` is NULL then `skbts` necessarily isn't and vice versa)?